CISCO BGP4 MIB support

sudharr86 opened the pull request:
### CISCO-BGP4-MIB support for FRRouting

This adds full support for the CISCO-BGP4-MIB (OID 1.3.6.1.4.1.9.9.187) in bgpd.
The Cisco MIB is widely used in production networks and provides much richer BGP
peer and route telemetry than the standard BGP4-MIB.

Reference: https://github.com/cisco/cisco-mibs/blob/main/v2/CISCO-BGP4-MIB.my

### What's implemented

**Global scalars** — cbgpLocalAs and cbgpNotifsEnable

**Peer tables** — cbgpPeerTable (legacy IPv4), cbgpPeer2Table (IPv4+IPv6),
cbgpPeer3Table (VRF-aware). Covers peer state, timers, connection details,
FSM transitions, error codes, previous state, and last error text.

**Capability tables** — cbgpPeerCapsTable and cbgpPeer2CapsTable.
Capabilities are reconstructed on-demand from parsed peer state,
so there's zero per-peer memory overhead. Supports MP, Route Refresh,
Extended Nexthop, Extended Message, Role, Graceful Restart, 4-byte AS,
Dynamic, Addpath, Enhanced Route Refresh, LLGR, FQDN, and legacy
Route Refresh (code 128).

**Address-family prefix tables** — Per-AFI/SAFI prefix counters including
accepted, denied, advertised, withdrawn, and suppressed counts.
Also exposes prefix admin limit, threshold, clear-threshold, and
MinASOriginationInterval.

**Route table** — cbgpRouteTable with Loc-RIB entries for IPv4 unicast,
IPv6 unicast, and L2VPN EVPN. Per-route attributes include origin,
AS-path, next-hop, MED, local-pref, aggregator info, best-path flag,
and unknown/transit attributes.

**Trap notifications** — All 10 notification types:
- Legacy IPv4: FSM state change, backward transition, prefix threshold exceeded/clear
- Peer2 (IPv4/IPv6): established, backward transition, FSM state change,
  backward transition (extended), prefix threshold exceeded/clear

### Infrastructure additions

- Per-peer filtered/withdrawn/suppressed prefix counters with hooks for
  prefix threshold events
- Peer list sorted by IP address for correct SNMP GetNext lexicographic ordering

### Commits

1. **bgpd: add filtered prefix tracking infrastructure** — per-peer/per-AFI counters
   for filtered, withdrawn, and suppressed prefixes; hook points for threshold events
2. **bgpd: add peer list sorted by address for SNMP** — maintains address-sorted peer
   list so MIB walks return OIDs in the right order
3. **bgpd: implement CISCO-BGP4-MIB support** — the main implementation (~4500 lines),
   all tables, traps, and capability reconstruction
4. **tests: add topotest for CISCO-BGP4-MIB** — comprehensive test covering SNMP walks,
   per-OID validation, SNMP-vs-CLI cross-checks, trap verification, filtered prefix
   counting, and no-peers edge case
5. **doc: add CISCO-BGP4-MIB documentation** — user-facing docs with configuration
   guidance and snmpwalk examples

### Testing

Topotest results: **7 passed, 1 skipped** (memory leak test disabled — standard infra skip).
Trap tests pass with snmptrapd installed.

donaldsharp commented on bgpd/bgp_snmp.c:
```diff
@@ -154,6 +156,196 @@ static int bgp_snmp_module_init(void)
+	 * the exact peer has disappeared between requests.
+	 */
+	for (ALL_LIST_ELEMENTS_RO(bgp->peer_by_addr, node, peer)) {
+		if (!peer->connection)
```
how do we not have a peer->connection?

sudharr86 commented on bgpd/bgp_snmp.c:
```diff
@@ -154,6 +156,196 @@ static int bgp_snmp_module_init(void)
+	/*
+	 * Iterate through the peers in the VRF
+	 */
+	if (!bgp->peer_by_addr)
```
dynamic peers, but we haven't learnt any peers.

donaldsharp replied:
actually strike that.  Why do we need *yet* another list of peers that BGP has?  We already store the peers in a hash table.  Additionally a linked list?  <sigh>. We can potentially have several thousand peers that we are going to be paying this price for addition/removal.  Not happy

sudharr86 commented on bgpd/bgp_snmp.c:
```diff
@@ -154,6 +156,196 @@ static int bgp_snmp_module_init(void)
+
+	if (bgp) {
+		/* Lookup in VRF */
+		peer = bgp_snmp_lookup_peer_in_vrf(bgp, addr);
```
agreed.

donaldsharp commented on bgpd/bgp_snmp.c:
```diff
@@ -154,6 +156,196 @@ static int bgp_snmp_module_init(void)
+	struct listnode *node;
+	struct listnode *head;
+
+	if (!bgp || !bgp->peer_by_addr)
```
again this function is called in several places where the bgp pointer is already checked or we know it cannot be null.  Let's pick one place to do it and stick with that

donaldsharp commented on bgpd/bgpd.c:
```diff
@@ -3614,15 +3688,18 @@ static struct bgp *bgp_create(as_t *as, const char *name,
 	SET_FLAG(bgp->peer_self->cap, PEER_CAP_AS4_ADV);
 
 	bgp->peer = list_new();
+	bgp->peer_by_addr = list_new();
```
ok I'm back to the question, why are we ever checking for the existence of this pointer?

donaldsharp commented on bgpd/bgpd.c:
```diff
@@ -2876,15 +2942,23 @@ int peer_delete(struct peer *peer)
 	if (!CHECK_FLAG(peer->sflags, PEER_STATUS_GROUP)
 	    && (pn = listnode_lookup(bgp->peer, peer))) {
 		/*
-		 * Removing from the list node first because
```
Why the change here?

donaldsharp commented on bgpd/bgpd.c:
```diff
@@ -1004,6 +1032,42 @@ int peer_cmp(struct peer *p1, struct peer *p2)
 	return sockunion_cmp(&p1->connection->su, &p2->connection->su);
 }
 
+/* Peer comparison function for sorting.  */
```
If we are actually doing this, this function *MUST* *MUST* handled the unnumbered case where it doesn't result in a pointer comparison.  You will get indeterminate results between runs.  Additionally I see nothing in here for the case where the v6 LL address is the same but the ifindex is different.  It will not work correctly at all.

sudharr86 replied:
agreed. BUT SNMP MIB walks for cisco mib cannot support same LL addresses, as SNMP walks should have increasing OIDs which is not possible with same LL addresses. Have fixed the code to allow peers with same LL addresses to be in the list. but the walk will return only the first one.

donaldsharp commented:
please update the documentation to point out that we now support this mib.  There are more than a few places that this will need to be addressed.  Look for other places in the documentation where mibs are talked about and update those too.

donaldsharp commented:
I've run out of time this evening to continue looking.  I would expect that there is going to need to be more than a few rounds of this given the size and non-broken up commits.

sudharr86 commented on bgpd/bgp_route.c:
```diff
@@ -4648,6 +4647,8 @@ static uint32_t bgp_filtered_routes_count(struct peer *peer, afi_t afi,
+		filtered = false;
+
 		for (ain = dest->adj_in; ain; ain = ain->next) {
 			const struct prefix *rn_p = bgp_dest_get_prefix(dest);
```
rewrote the logic of calculating filtered prefix count.

sudharr86 commented on bgpd/bgp_snmp.c:
```diff
@@ -154,6 +156,196 @@ static int bgp_snmp_module_init(void)
+		sa_family_t fam;
+		struct ipaddr peer_addr;
+
+		if (!peer->connection)
```
fixed.

sudharr86 commented on bgpd/bgp_snmp_cbgp4.c:
```diff
@@ -0,0 +1,1847 @@
+	bool addr_present = false;
+	sa_family_t family = AF_UNSPEC;
+
+	/* Somehow with net-snmp 5.7.3, every OID item in an array
```
fixed.

ton31337 commented on tests/topotests/bgp_snmp_cbgp4mib/test_bgp_snmp_cbgp4mib.py:
```diff
@@ -0,0 +1,916 @@
+#!/usr/bin/env python
+#
+# Copyright (c) 2023 Cisco and/or its affiliates.
```
2023?

ton31337 commented on tests/topotests/bgp_snmp_cbgp4mib/test_bgp_snmp_cbgp4mib.py:
```diff
@@ -0,0 +1,916 @@
+#!/usr/bin/env python
```
Please add SPDX-License tag.

sudharr86 commented on bgpd/bgpd.h:
```diff
@@ -509,6 +509,7 @@ struct bgp {
 
 	/* BGP peer. */
 	struct list *peer;
+	struct list *peer_by_addr;
```
I had thought about this first. I did not like implementing the new peer list as well.  But it wasn't that simple. Here is my reasoning. 

We CANNOT use an ordered-by-address list by default because:
- Peer groups and peers are mixed in the same list
- Would break user-visible output (show commands, config)
- Would change configuration file ordering
- Current ordering has semantic meaning (groups first, interfaces prioritized)
- SNMP needs pure address ordering (VRF→AF→Address), but BGP needs logical grouping

The separate peer_by_addr list is necessary because:
- SNMP requires strict OID-based ordering (AF type, then address)
- BGP operations require logical grouping (peer-groups, interfaces, then addresses)
- These two requirements are fundamentally incompatible
- The overhead is minimal (just list node pointers, ~16 bytes per peer)
- 

To use one list would require either:
- Breaking existing BGP CLI/config behavior, OR
- Making SNMP walk through an incorrectly-ordered list (breaking SNMP). 
- Do a full walk of the peer list to get the next peer for each snmp walk, which is not good for scale.

So the dual-list approach seems to be better  solution.

sudharr86 commented on bgpd/bgp_route.c:
```diff
@@ -5395,6 +5361,10 @@ void bgp_update(struct peer *peer, const struct prefix *p, uint32_t addpath_id,
 		peer->stat_pfx_filter++;
+		if (ain && !ain->filtered) {
+			peer->pfiltered[afi][orig_safi]++;
+			ain->filtered = 1;
```
agreed.

donaldsharp commented on doc/user/snmp.rst:
```diff
@@ -122,6 +138,32 @@ An example below is how to query SNMP for BGP:
       $ # BGP4-MIB (https://www.circitor.fr/Mibs/Mib/B/BGP4-MIB.mib)
       $ snmpwalk -c public -v2c -On -Ln localhost .1.3.6.1.2.1.15
 
+      $ # CISCO-BGP4-MIB (https://www.circitor.fr/Mibs/Mib/C/CISCO-BGP4-MIB.mib)
```
This link does not appear to do anything for me.  It must be fixed

ton31337 commented on bgpd/bgpd.c:
```diff
@@ -2093,6 +2155,20 @@ struct peer *peer_create(union sockunion *su, const char *conf_if,
 	listnode_add_sort(bgp->peer, peer);
+	peer = peer_lock(peer);
+	/* Add to peer_by_addr list in sorted order */
+	{
```
Any reason we use a separate block here?

ton31337 reviewed (changes requested):
Please review the topotest to make sure we don't use too low timers (wait/count). 15 seconds is the minimum.

I'm curious what we gain/loose if we just do list_sort() when SNMP is needed without introducing peer_by_addr (another sorted list)?

riw777 commented on bgpd/bgp_advertise.c:
```diff
@@ -162,10 +162,10 @@ bool bgp_adj_out_lookup(struct peer *peer, struct bgp_dest *dest,
-void bgp_adj_in_set(struct bgp_dest *dest, struct peer *peer, struct attr *attr,
-		    uint32_t addpath_id, struct bgp_labels *labels)
+struct bgp_adj_in *bgp_adj_in_set(struct bgp_dest *dest, struct peer *peer, struct attr *attr,
+				  uint32_t addpath_id, struct bgp_labels *labels)
```
not certain this white space change is needed?

riw777 commented on bgpd/bgp_snmp_cbgp4.h:
```diff
@@ -0,0 +1,215 @@
+#define CBGP4MIB 1, 3, 6, 1, 4, 1, 9, 9, 187
+
+/* ciscoBgp4MIB Notifications */
+#define CBGP_NOTIFY_PREFIX		       0
```
the alignment for these values look pretty random to me?

riw777 reviewed (commented):
just a few questions ... nothing major

sudharr86 committed:
bgpd: add filtered prefix tracking infrastructure

Add per-peer, per-AFI/SAFI tracking of filtered (denied) prefixes,
withdrawn prefixes, and suppressed (damped) prefixes. The pfiltered
counter increments at the centralized filtered: label in
bgp_update() so that every deny path (ACL, prefix-list, route-map,
ORF, etc.) is captured.

Also adds hook infrastructure for prefix threshold exceeded/clear
events. New counters are reset on peer session teardown in bgp_stop().
Updated show commands (JSON and text) to display the counters.

Signed-off-by: Sudharsan Rajagopalan <sudhar86work@gmail.com>

sudharr86 committed:
bgpd: add peer list sorted by address for SNMP

Maintain a separate peer list sorted by IP address (IPv4 before
IPv6, numerically ordered within each family) so that SNMP
GetNext walks return peers in the lexicographic order required
by the MIB index encoding.

Signed-off-by: Sudharsan Rajagopalan <sudhar86work@gmail.com>

sudharr86 committed:
bgpd: implement CISCO-BGP4-MIB support

Implement the CISCO-BGP4-MIB (OID 1.3.6.1.4.1.9.9.187) for FRR's
BGP daemon, providing SNMP management of BGP peers, address families,
prefix counters, capabilities, and route table entries across
cbgpPeer, cbgpPeer2, and cbgpPeer3 table hierarchies.

CISCO-BGP4-MIB trap notifications are gated behind the
"bgp snmp traps cisco-bgp4" configuration command (disabled by
default) to avoid interfering with existing BGP4-MIB and
BGP4V2-MIB trap receivers.

Signed-off-by: Sudharsan Rajagopalan <sudhar86work@gmail.com>

sudharr86 committed:
tests: add topotest for CISCO-BGP4-MIB

Add comprehensive topotest for CISCO-BGP4-MIB covering cbgpPeer,
cbgpPeer2, cbgpPeer3, cbgpRouteTable, prefix counters, trap
notifications, SNMP-vs-CLI cross-validation, and no-peers crash
regression.

Signed-off-by: Sudharsan Rajagopalan <sudhar86work@gmail.com>

sudharr86 committed:
doc: add CISCO-BGP4-MIB documentation

Document CISCO-BGP4-MIB support including peer tables (cbgpPeer,
cbgpPeer2, cbgpPeer3), route table with per-AFI/SAFI statistics,
and trap notification configuration.

Signed-off-by: Sudharsan Rajagopalan <sudhar86work@gmail.com>

riw777 commented:
@ton31337 can you re-reveiew this to see if we can clear the blocker?

riw777 reviewed (approved):
looks good
